File: Arrays/Find_Missing_No/better_sol.py
def missingNum(arr):
    n = len(arr)+1  #(1 missing no)
    hash = [0]*(n+1)
    for num in arr:
        hash[num] += 1

    for i in range(1, n + 1):
        if hash[i] == 0:
            return i

    
    return -1

if __name__ == '__main__':
    arr = [8, 2, 4, 5, 3, 7, 1]
    res = missingNum(arr)
    print(res)

File: Arrays/Find_Missing_No/better_sol.cpp
#include <iostream>
#include <vector>
using namespace std;

int missingNum(vector<int> &arr){
    int n = arr.size()+1;

    //create a hash array of size n
    vector<int> hash(n+1,0);

    //store fequencies of element
    for(int i=0; i<n-1; i++){
        hash[arr[i]]++;
    }

    // find the missing no
    for(int i=1; i<=n;i++){
        if (hash[i] == 0){
            return i;
        } 

    }

    return -1;
}


int main() {
    vector<int> arr = {8, 2, 4, 5, 3, 7, 1};
    int res = missingNum(arr);
    cout << res << endl;
    return 0;
}